lib/blocks.php

```diff
@@ -296,3 +296,209 @@ function gutenberg_enqueue_editor_block_styles_assets() {
 	}
 	add_action( 'enqueue_block_editor_assets', 'gutenberg_enqueue_editor_block_styles_assets' );
 }
+
+/**
+ * Renders the classNames and styles for blocks
+ *
+ * @param string $block_content Output of the current block.
+ * @param array  $block Block Object.
+ * @return string New block output.
+ */
+function gutenberg_experimental_apply_classnames_and_styles( $block_content, $block ) {
+	if ( ! isset( $block['attrs'] ) ) {
+		return $block_content;
+	}
+
+	// Check what style features the block supports.
+	$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+	if ( ! $block_type ) {
+		return $block_content;
+	}
+
+	$supports = gutenberg_experimental_global_styles_get_supported_styles( $block_type->supports );
+
+	// Return early if nothing is supported.
+	if ( 0 === count( $supports ) ) {
+		return $block_content;
+	}
+
+	$colors     = gutenberg_experimental_build_css_colors( $block['attrs'], $supports );
+	$typography = gutenberg_experimental_build_css_typography( $block['attrs'], $supports );
+
+	$extra_classes = array_merge(
+		$colors['css_classes'],
+		$typography['css_classes'],
+		isset( $block['attrs']['className'] ) ? array( $block['attrs']['className'] ) : array(),
+		isset( $block['attrs']['align'] ) ? array( 'has-text-align-' . $block['attrs']['align'] ) : array()
+	);
+	$extra_styles  = (
+		$colors['inline_styles'] ||
+		$typography['inline_styles']
+	) ? esc_attr( $colors['inline_styles'] ) .
+		esc_attr( $typography['inline_styles'] )
+	: '';
+
+	$dom = new DOMDocument( '1.0', 'utf-8' );
+
+	// Suppress warnings from this method from polluting the front-end.
+	// @codingStandardsIgnoreStart
+	if ( ! @$dom->loadHTML( $block_content, LIBXML_HTML_NOIMPLIED | LIBXML_HTML_NODEFDTD | LIBXML_COMPACT ) ) {
+	// @codingStandardsIgnoreEnd
+		return $block_content;
+	}
+
+	$xpath      = new DOMXPath( $dom );
```

Comment:
I wonder if regex is more performant. This is definitely stronger and safer but in the long run, performance is important for the frontend. I'm happy starting with this and monitor impact

Reply:
I had a similar thought about performance, but attempting to parse HTML with a regex gets messy really quickly.

FWIW, I found myself in a similar situation a while ago, when [adding a new sniff](https://github.com/WordPress/WordPress-Coding-Standards/pull/1856) to WPCS. I ended up using [XMLReader](https://www.php.net/manual/en/book.xmlreader.php) which out of a number of options seemed to have the smallest resource footprint. However, I only needed _parsing_ at the time -- I didn't have to _write_ HTML.

For writing, there's however [XMLWriter](https://www.php.net/manual/de/book.xmlwriter.php) -- maybe it could be used in combination with XMLReader.

[Here's](https://stackoverflow.com/questions/3577641/how-do-you-parse-and-process-html-xml-in-php) a StackOverflow post listing PHP's built-in (and external) HTML parsing options.

[Here's](https://github.com/WordPress/WordPress-Coding-Standards/pull/1856#discussion_r373473501) some more background on RegEx-based parsing (and why it's messy) and alternatives.